include/TriangleNodesProvisioner.hpp

```diff
@@ -14,15 +14,40 @@
 #include <memory>
 
 namespace blitzdg {
+    using real_mat_smart_ptr = std::unique_ptr<real_matrix_type>;
+    using real_vec_smart_ptr = std::unique_ptr<real_vector_type>;
+
+    struct DGContext2D {
+        const index_type& Np;
+        const index_type& Nfp;
+        const index_type& K;
+        const index_type& NumFaces;
+        const real_matrix_type& Filt;
+        const real_matrix_type& x;
+        const real_matrix_type& y;
+        const real_matrix_type& Fscale;
+        const real_matrix_type& J;
+        const real_matrix_type& rx;
+        const real_matrix_type& sx;
+        const real_matrix_type& ry;
+        const real_matrix_type& sy;
+        const real_matrix_type& nx;
+        const real_matrix_type& ny;
+        const real_matrix_type& Dr;
+        const real_matrix_type& Ds;
+        const index_hashmap& bcHash;
+        const index_vector_type& vmapM;
+        const index_vector_type& vmapP;
+        const real_matrix_type& Lift;
+    };
```

Comment:
Declaring the members of this struct as const references is a bit odd. Doing so means that you can't assign to a `DGContext2D` and all member values must be specified at initialization. Furthermore, if `DGConext2D` is a member of a class, then that class is no-longer assignable. If these are the semantics that you want, then it's fine. If not, then I would consider having it hold non-owning (raw) pointers to the given members. If you don't want the objects pointed to by the member pointers to be modified, then use const pointers.

Reply:
I've gone back-and-forth with this one. I agree from a raw implementation point of view, so I started switching things over to use the const raw-pointers then realized I would need a dereference operator almost everywhere there was a dg.(something) in the sw2d/main.cpp code. 

My issue is that the main 'solver' codes are supposed to be penetrable by somebody who isn't too C++ savvy. I'm worried about scaring them off with all the pointer semantics.

I think I'll leave it as a read-only context with const references as is. I don't think these references need to be changed after the TriangleNodesProvisioner is created as far as I can think. The Provisioner being unassignable isn't amazing, but I can't think of too many cases where we need assignments, such cases are fairly esoteric, but if they become common I'll reconsider the choice I'm making here.

Reply:
I would caution you about using `const` members as this is rarely what you want. Since you return the context by const reference, those member references will already be non-mutable, so you don't need the const on them. Another design you could consider would be to declare `DGContext2D` as a class with private member variables stored as raw pointers and then simply add public members that return the scalars by value and the pointed to members by const reference. Since you goal is to only provide read-access to these variables, this would also be a good approach.